### PR TITLE
Dataset prio 1 (consolidated + unconsolidated)

### LINKED ISSUE #490: Prio 1 datasets (consolidated and unconsolidated)
Follow up of #427

Include all Prio 1 labels for both consolidated and unconsolidated rock (see https://github.com/swisstopo/swissgeol-boreholes-dataextraction/issues/390, https://github.com/swisstopo/swissgeol-boreholes-dataextraction/issues/391, https://github.com/swisstopo/swissgeol-boreholes-dataextraction/issues/409, https://github.com/swisstopo/swissgeol-boreholes-dataextraction/issues/410, https://github.com/swisstopo/swissgeol-boreholes-dataextraction/issues/411, https://github.com/swisstopo/swissgeol-boreholes-dataextraction/issues/412, https://github.com/swisstopo/swissgeol-boreholes-dataextraction/issues/413, https://github.com/swisstopo/swissgeol-boreholes-dataextraction/issues/414, https://github.com/swisstopo/swissgeol-boreholes-dataextraction/issues/415) as datasets

### PR DESCRIPTION
**Pull request by christianabbet**
Closes #490 

# Data Overview

| Dataset               | Support (num classes) | Target | Issue | BERT* |
|-----------------------|-----------------------|--------|-------|-------|
| Accessory components  | 75 (66)	            | Multi  | #410  |   NO  |
| Alteration degree (c) | 2,716 (8)             | Single | #414  |  YES+ |
| Alteration degree (u) | 41 (8)                | Single | #414  |   NO+ |
| Cementation           | 78 (7)                | Single | #415  |   NO  |
| Color (c)             | 16,143 (91)           | Single | #390  |  YES+ |
| Color	(u)             | 20,121 (91)           | Single | #390  |  YES+ |
| Debris                | 70,084, (7)           | Multi  | #412  |  YES  |
| EN Main               | 89,842 (34)           | Single | -     |  YES  |
| Grain angularity      | 70,377 (8)            | Multi  | #391  |  YES  |
| Grain shape           | 70,087 (5)            | Multi  | #413  |  YES  |
| Lithology             | 45,323 (61)           | Single | -     |  YES  |
| Mineral components    | 63 (111)	            | Multi  | #409  |   NO  |
| Organic components    | 70,102 (11)           | Multi  | #411  |  YES  |
| USCS                  | 9,917 (38)            | Single | -     |  YES  |

> (*) Indicate if BERT training is possible considering amount of data.
(+) Training possible with BERT including consolidated and unconsolidated

Class ordering is now consistent: "not specified" is always first, followed by all dataset classes in alphabetical order.

## Lithology data leakage fix

Updated lithology class names to match the current ground truth: "tuffit" → "tuff", "ka" → "not specified", and added the "other" label. Also corrected "cataclastite" → "cataclasite" and "unconsolidated deposits" → "unconsolidated" in ground truth. 

The legacy ground truth file in data/lithology_splits has been refactored to match the new structure. Test/val/train splits were audited and fixed to remove entries sourced from the same document (data leakage).

### CODE REVIEW

File: docs/train_BERT.md [hunk 1]
```diff
@@ -11,11 +11,21 @@ Each file contains boreholes, and each borehole contains layers. For each layer:
 
 | Class | Config string | JSON layer tag |
 |-------|--------------|----------------|
+| `AccessoryComponentsSystem` | `accessory_components` | `consolidated.accessory_components` |
```
Comment (AgathaSchmidt) [inline docs/train_BERT.md:14]: when does the ClassificatioSystem name include "(un)consolidated" and when not? 
For sure it  has to be there when we have both like in Color and AlterationDegree but for the other ones sometimes its there, e.g. "DebrisUnconsolidatedSystem" , "OrganicComponentsUnconsolidatedSystem" . but not fall all unconsolidated e.g. "GrainAngularitySystem" and for no with consolidated

File: src/classification/config/bert/bert_config_alteration_degree_a2a.yml [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+  global_consolidated:
+    classification_system: alteration_degree_consolidated
+    ground_truths:
+      - geoquat_ground_truth.json
+      - nagra_ground_truth.json
+  geoquat_consolidated:
+    classification_system: alteration_degree_consolidated
+    ground_truths:
+      - geoquat_ground_truth.json
+  nagra_consolidated:
```
Comment (AgathaSchmidt) [inline src/classification/config/bert/bert_config_alteration_degree_a2a.yml:31]: do you think it will be a problem/error, that nagra has no data for consolidated ? --> empty test set in this case
Only unconsolidated

File: src/classification/models/train.py [hunk 3]
```diff
@@ -391,30 +414,57 @@ def __init__(self, id2class_enum: dict):
...
         logits, labels = eval_pred
-        predictions = np.argmax(logits, axis=-1)
-
-        self._cm = confusion_matrix(labels, predictions, labels=self._sorted_ids)
-        per_class_results = per_class_metric(predictions, labels)
-        class_metrics = {cls: per_class_results.get(id_cls, Metrics()) for id_cls, cls in self._id2class_enum.items()}
-        per_class_metrics = {f"{cls.name}_f1": metric.f1 for cls, metric in class_metrics.items()}
-        overall_metrics = Metrics.micro_average(class_metrics.values()).to_dict("all_micro")
-        return overall_metrics | per_class_metrics
+        if labels.ndim == 2:  # multi-label
```
Comment (AgathaSchmidt) [inline src/classification/models/train.py]: maybe we should push the multi class PR to develop first and then you pull the changes an resolve potential conflicts

File: src/classification/utils/datasets/grain_shape.py [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+class GrainShapeSystem(ClassificationSystem):
+    """Classification system for grain shape of unconsolidated geological layers."""
+
+    @classmethod
+    def normalize_class_string(cls, class_str: str) -> str:
+        """Normalize a grain shape class string.
+
+        Args:
+            class_str (str): The class string to be normalized (e.g. "platy").
```
Comment (AgathaSchmidt) [inline src/classification/utils/datasets/grain_shape.py]: use an example which is actually normalized , e.g. "not specified"

File: src/classification/utils/datasets/grain_shape.py [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+
+        Returns:
+            str: The normalized grain shape class string.
```
Comment (AgathaSchmidt) [inline src/classification/utils/datasets/grain_shape.py]: add normalized example "not_specified"

File: src/classification/utils/datasets/lithology.py [hunk 6]
```diff
@@ -17,23 +17,27 @@ class LithologySystem(ClassificationSystem):
...
     def normalize_class_string(cls, class_str: str) -> str:
         """Normalize a lithology class string.
 
-        Examples include test that are comma separated (e.g. "limestone, bioclasts")
-            or semicolon ("limestone: micritic")
+        Examples include texts that are semicolon/comma separated (e.g. "limestone: micritic" -> "limestone",
+        "sandstone, marly" -> "sandstone") or white space (e.g. "not specified" -> "not_specified")
 
         Args:
-            class_str (str): The class string to be normalized (e.g. " "limestone, bioclasts",").
```
Comment (AgathaSchmidt) [inline src/classification/utils/datasets/lithology.py:28]: why did you specifically remove the example? 
We have them for every other task
Also, this one is more complicated than the other ones, so an example could be especially useful in my opinion

File: src/classification/utils/datasets/mineral_components.py [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+
+class MineralComponentsSystem(ClassificationSystem):
+    """Classification system for mineral components of consolidated geological layers."""
+
+    @classmethod
+    def normalize_class_string(cls, class_str: str) -> str:
+        """Normalize a mineral class string.
+
+        Args:
+            class_str (str): The class string to be normalized (e.g. "pyrite").
```
Comment (AgathaSchmidt) [inline src/classification/utils/datasets/mineral_components.py]: Maybe we should use an example where the normalization is actually doing something, e.g. "not specified" --> "not_specified"

File: src/classification/utils/datasets/uscs.py [hunk 8]
```diff
@@ -59,42 +59,41 @@ class USCSClasses(IntEnum):
             at 0, so using 0-based indexing avoids the need to address the issue later and prevents potential bugs.
         """
 
-        kunst = 0
+        kA = 0
```
Comment (AgathaSchmidt) [inline src/classification/utils/datasets/uscs.py]: In lithology we changed "kA" to "not specified" 
Should we do the same thing here?

File: src/classification/utils/datasets/classification.py
Comment (AgathaSchmidt) [inline src/classification/utils/datasets/classification.py]: due to the change in 182 where `ground_truth_class` was changed to `None | list[ClassificationSystem.EnumMember] ` 

we need to change this line as well 

 ```
           "ground_truth_class": [c.name for c in self.ground_truth_class]
            if self.ground_truth_class is not None
            else None,
```

File: src/classification/utils/data_utils.py
Comment (AgathaSchmidt) [inline src/classification/utils/data_utils.py:134]: Also, because `ground_truth_class` is a list now
`get_first_key_class=lambda layer: layer.ground_truth_class[0] if layer.ground_truth_class else None,`

Comment (AgathaSchmidt) [inline src/classification/utils/data_utils.py:149]: `get_second_key_class=lambda layer: layer.ground_truth_class[0] if layer.ground_truth_class else None,`

Comment (AgathaSchmidt) [inline src/classification/utils/data_utils.py]: `all_classes = set(predictions) | {cls for gt in ground_truths if gt for cls in gt}`

Comment (AgathaSchmidt) [inline src/classification/utils/data_utils.py]: `num_ground_truth = sum(class_ in gt for gt in ground_truths if gt)`

Comment (AgathaSchmidt) [inline src/classification/utils/data_utils.py]: `"ground_truth_class": [c.name for c in layer.ground_truth_class] if layer.ground_truth_class else None,`

**Review by AgathaSchmidt — CHANGES_REQUESTED** [review]
I tested the bert training pipeline on `organic_components` (multi) and `cementation` (single) and it works fine. 

Mostly few consistency remarks.  

When running the classification pipeline I encountered several errors which came due to changing the ground_truth_class to a list 

This one has to be corrected as well: 


in `src/classififcation/evaluation/evaluate.py `

 change line 226 

`   labels = [layer.ground_truth_class for layer in layers]`

to 

` labels = [layer.ground_truth_class[0] if layer.ground_truth_class else None for layer in layers]
`

**Comment by AgathaSchmidt** [comment]
Also, we need to change line 39 `in src/classification/classifiers/bert_classifier.py`

from 
`model_path = self.config["model_path"]`
to 
` model_path = Path(self.config["model_path"])`

File: docs/train_BERT.md (hunk 1 above)
Comment (christianabbet) [reply docs/train_BERT.md:14]: Good point I missed the two. I'll put conso/unconso only if both exists (color and alteration)

File: src/classification/config/bert/bert_config_alteration_degree_a2a.yml (hunk 2 above)
Comment (christianabbet) [reply src/classification/config/bert/bert_config_alteration_degree_a2a.yml:31]: I was not sure. I kept the full schema even if it was empty such that we have the same logic everytime. We just have the explicit message that it's empty. In the future if we augment the datasets then we don't need to modify it

File: src/classification/models/train.py (hunk 3 above)
Comment (christianabbet) [reply src/classification/models/train.py]: Agree! I'll wait for the other PR

File: src/classification/utils/datasets/lithology.py (hunk 6 above)
Comment (christianabbet) [reply src/classification/utils/datasets/lithology.py:28]: I put the examples above. Do you think we should add more ?

Comment (AgathaSchmidt) [reply src/classification/utils/datasets/lithology.py:28]: Ah, sorry, no its fine like that! 
I was looking at the wrong place

**Review by AgathaSchmidt — APPROVED** [review]
Looks good
I tested classification, extraction and Bert training pipeline (single and multi label) , all worked fine
